src/Sfx/App/Scripts/Directives/RingViewDirective.ts

```diff
@@ -0,0 +1,231 @@
+        });
+
+        let candidateEndpoints = [];
+
+        m.find("Node").each((idx, n) => {
+          candidateEndpoints.push($(n).attr("IPAddressOrFQDN") + ":" + tmap[$(n).attr("NodeTypeRef")]);
+        })
+
+        // TODO load from config
+        ws = new WebSocket("ws://" + candidateEndpoints[0]);
+        // ws = new WebSocket("ws://127.0.0.1:10286");
+  
+        ws.onerror = function (error) {
+          console.log(error)
+          if (recreateTimer){
+            clearTimeout(recreateTimer);
+          }
+          recreateTimer = setTimeout(recreateWs, 1000);
+        };
+
+        ws.onmessage = function (message) {
+          let json;
+          try {
+            json = JSON.parse(message.data);
+          } catch (e) {
+            console.log('This doesn\'t look like a valid JSON: ', message.data);
+            return;
+          }
+          $scope.messageHandler(json);
+        };
+
+        ws.onopen = function(){
+          candidateEndpoints.forEach((ip) => {
+            $scope.sendQuery(ip);
+          })
+        };
+      }
+
+      $scope.clusterManifest.ensureInitialized().then( data => {
+        recreateWs();
+      })
+
+      $scope.sendQuery = function(endpint: string){
+        if(ws.readyState !== WebSocket.OPEN){
+          return;
+        }
+
+        ws.send(JSON.stringify(
+          {
+            "message_type": "query",
+            "address": endpint,
+          }
+        ))
+      }
+
+      $scope.addNode = function(){
+        $scope.sendQuery($scope.ipaddr);
+        $scope.ipaddr = "";
+      }
+
+      let cy = cytoscape({
+        container: document.getElementById('cytoscape-canvas'),
+        zoomingEnabled: false,
+        userZoomingEnabled: false,
+        autoungrabify: false,
+        style: [{
+          "selector": "node",
+          "style": {
+              "content": "data(label)",
+              "font-size": "12px",
+              "text-valign": "center",
+              "text-halign": "center",
+              "background-color": "#7FBA00",
+              "text-outline-color": "#555",
+              "text-outline-width": "2px",
+              "color": "#fff",
+              "overlay-padding": "6px",
+              "z-index": "10"
+          }
+       }],
+      });
+
+      let hideAllTippies = function(){
+        cy.nodes().forEach(hideTippy);
+      };
+
+      cy.on('tap', function(e){
+        if(e.target === cy){
+          hideAllTippies();
+        }
+      });
+  
+      cy.on('tap', 'edge', function(e){
+        hideAllTippies();
+      });
+  
+      cy.on('zoom pan', function(e){
+        hideAllTippies();
+      });   
+
+      $(".main-view").scroll(function(e){
+        hideAllTippies();
+      });
+
+      $scope.buildLabel = function(node: any){
+        // return node.node_name + " (" + node.phase + ")";
+        return node.node_id.substring(0, 5) + " (" + node.phase + ")";
+      }
+
+      $scope.messageHandler = function (node: any) {
```

Comment:
can we add a node interface or if this is a reference to a node class that already exists in the project can we use that.

Reply:
can we use `any`? it is returned by server side json

Reply:
where exactly is this getting returned from? Is there a spec for what the json looks like?